File: .claude/hooks/python/validate-commit.py
#!/usr/bin/env python3
"""
Hook: PreToolCall — Bash
Chặn git commit với message sai format.
Generic version — hoạt động với mọi project dùng Conventional Commits.
Customize COMMIT_TYPES và MAX_SUBJECT_LEN cho project của bạn.

Exit 0 = cho phép Claude tiếp tục
Exit 1 = block Claude, hiển thị error
"""
import sys, json, re, io

if sys.stdout.encoding and sys.stdout.encoding.lower() != 'utf-8':
    try:
        sys.stdout = io.TextIOWrapper(sys.stdout.buffer, encoding='utf-8', errors='replace')
    except Exception:
        pass

# ── Config — sửa theo project ─────────────────────────────────────────────────
COMMIT_TYPES    = "feat|fix|hotfix|refactor|test|docs|ci|chore|perf|security"
MAX_SUBJECT_LEN = 72
MAX_TOTAL_LEN   = 500
BAD_MESSAGES    = [
    r"^fix\s+bug", r"^fixed$", r"^update\s+code", r"^update$",
    r"^wip", r"^temp", r"^test\s+commit", r"^minor$",
    r"^changes$", r"^\.\.*$", r"^sửa", r"^fix\.$",
]

try:
    data    = json.loads(sys.stdin.read() or "{}")
    command = data.get("tool_input", {}).get("command", "")
except (json.JSONDecodeError, ValueError):
    sys.exit(0)

if "git commit" not in command:
    sys.exit(0)

msg = ""
for pat in [r'-m\s+"((?:[^"\\]|\\.)+)"', r"-m\s+'((?:[^'\\]|\\.)+)'"]:
    m = re.search(pat, command)
    if m:
        msg = m.group(1).replace('\\"', '"').replace("\\'", "'")
        break

if not msg:
    sys.exit(0)

subject = msg.split("\n")[0].strip()
errors, warnings = [], []

CC = re.compile(rf"^(\[AI\] )?({COMMIT_TYPES})(\([^)]+\))?!?: .+")
if not CC.match(subject):
    errors.append(
        f"❌  Commit message không đúng Conventional Commits format\n"
        f"    Got    : '{subject}'\n"
        f"    Expect : feat(scope): description  |  Types: {COMMIT_TYPES}"
    )

if len(subject) > MAX_SUBJECT_LEN:
    errors.append(f"❌  Subject dài quá ({len(subject)}/{MAX_SUBJECT_LEN} chars)")

if len(msg) > MAX_TOTAL_LEN:
    errors.append(f"❌  Message vượt {MAX_TOTAL_LEN} chars ({len(msg)} chars)")

for bad in BAD_MESSAGES:
    if re.match(bad, subject, re.IGNORECASE):
        errors.append(f"❌  Message quá chung chung: '{subject}'")
        break

has_bang   = bool(re.match(rf"^(\[AI\] )?({COMMIT_TYPES})(\([^)]+\))?!:", subject))
has_footer = "BREAKING CHANGE:" in msg
if has_bang and not has_footer:
    warnings.append(
        "⚠️   Breaking change (!) phát hiện — nên thêm:\n"
        "    BREAKING CHANGE: <mô tả thay đổi + migration plan>"
    )

D = "─" * 58
if errors:
    print(f"\n{D}\n🚫  COMMIT BLOCKED — Conventional Commits violation\n{D}")
    for e in errors:
        print(f"\n{e}")
    for w in warnings:
        print(f"\n{w}")
    print(f"\n{D}\n")
    sys.exit(1)

if warnings:
    print(f"\n{D}\n⚠️   COMMIT WARNING\n{D}")
    for w in warnings:
        print(f"\n{w}")
    print(f"\n{D}\n")

print(f"✅  Commit OK: '{subject}'")
sys.exit(0)

File: .claude/hooks/nodejs/validate-commit.js
#!/usr/bin/env node
// Hook: PreToolCall — Bash (Conventional Commits gate)
// Requires: Node.js >= 16 (zero dependencies)
'use strict';
const TYPES = 'feat|fix|hotfix|refactor|test|docs|ci|chore|perf|security';
const D = '─'.repeat(60);
let raw = '';
process.stdin.setEncoding('utf8');
process.stdin.on('data', c => { raw += c; });
process.stdin.on('end', () => {
  let data = {};
  try { data = JSON.parse(raw); } catch { process.exit(0); }
  const cmd = data?.tool_input?.command ?? '';
  if (!cmd.includes('git commit')) process.exit(0);

  let msg = '';
  const dq = cmd.match(/-m\s+"((?:[^"\\]|\\.)*)"/);
  const sq = cmd.match(/-m\s+'((?:[^'\\]|\\.)*)'/);
  if (dq) msg = dq[1].replace(/\\"/g, '"');
  else if (sq) msg = sq[1].replace(/\\'/g, "'");
  if (!msg) process.exit(0);

  const subject = msg.split('\n')[0].trim();
  const errors = [], warnings = [];

  const CC = new RegExp(`^(\\[AI\\] )?(${TYPES})(\\([^)]+\\))?!?: .+`);
  if (!CC.test(subject)) {
    errors.push(`❌  Wrong format: '${subject}'`);
    errors.push(`    Expected: feat(scope): desc  |  Types: ${TYPES}`);
  }
  if (subject.length > 72) errors.push(`❌  Subject too long (${subject.length}/72)`);
  if (msg.length > 500)    errors.push(`❌  Message exceeds 500 chars (${msg.length})`);
  if (/^(fix bug|fixed|update|wip|temp|minor|changes|\.+)$/i.test(subject))
    errors.push(`❌  Generic message: '${subject}'`);

  const bang = new RegExp(`^(\\[AI\\] )?(${TYPES})(\\([^)]+\\))?!:`).test(subject);
  if (bang && !msg.includes('BREAKING CHANGE:'))
    warnings.push('⚠️   Breaking change (!) — add BREAKING CHANGE: footer');

  if (errors.length) {
    console.log(`\n${D}\n🚫  COMMIT BLOCKED\n${D}`);
    errors.forEach(e => console.log(e));
    if (warnings.length) { console.log(''); warnings.forEach(w => console.log(w)); }
    console.log(`\n${D}\n`);
    process.exit(1);
  }
  if (warnings.length) {
    console.log(`\n${D}\n⚠️   WARNING\n${D}`);
    warnings.forEach(w => console.log(w));
    console.log(`\n${D}\n`);
  }
  console.log(`✅  Commit OK: '${subject}'`);
  process.exit(0);
});
